# Notebook Introduction

In this notebook I clean some data related to crimes committed between 2020 and 2024. The purpose of cleaning the data is to turn it into a format that can be used for basic EDA, and then imported into Tableau to create a visualisation.

The data includes information on the date and time the crime was committed, the geographical location, the type and severity of crime, and a few other features. The aim is to remove invalid data and ensure all columns are in the same format. I will also add and remove some columns depending on how important they will be for the visualisation.

# Importing the Data

I will start by importing the libraries and making sure the data is read in properly.

In [256]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from IPython.display import display, HTML

crime_df = pd.read_csv("Crime Data Original.csv")

In [257]:
crime_df.head()

DR_NO               Date Rptd                DATE OCC  TIME OCC  AREA  \
0  190326475  03/01/2020 12:00:00 AM  03/01/2020 12:00:00 AM      2130     7   
1  200106753  02/09/2020 12:00:00 AM  02/08/2020 12:00:00 AM      1800     1   
2  200320258  11/11/2020 12:00:00 AM  11/04/2020 12:00:00 AM      1700     3   
3  200907217  05/10/2023 12:00:00 AM  03/10/2020 12:00:00 AM      2037     9   
4  200412582  09/09/2020 12:00:00 AM  09/09/2020 12:00:00 AM       630     4   

    AREA NAME  Rpt Dist No  Part 1-2  Crm Cd  \
0    Wilshire          784         1     510   
1     Central          182         1     330   
2   Southwest          356         1     480   
3    Van Nuys          964         1     343   
4  Hollenbeck          413         1     510   

                                Crm Cd Desc  ... Status   Status Desc  \
0                          VEHICLE - STOLEN  ...     AA  Adult Arrest   
1                     BURGLARY FROM VEHICLE  ...     IC   Invest Cont   
2                             BIKE - STOLEN  ...     IC   Invest Cont   
3  SHOPLIFTING-GRAND THEFT ($950.01 & OVER)  ...     IC   Invest Cont   
4                          VEHICLE - STOLEN  ...     IC   Invest Cont   

  Crm Cd 1 Crm Cd 2  Crm Cd 3 Crm Cd 4  \
0    510.0    998.0       NaN      NaN   
1    330.0    998.0       NaN      NaN   
2    480.0      NaN       NaN      NaN   
3    343.0      NaN       NaN      NaN   
4    510.0      NaN       NaN      NaN   

                                   LOCATION Cross Street      LAT       LON  
0   1900 S  LONGWOOD                     AV          NaN  34.0375 -118.3506  
1   1000 S  FLOWER                       ST          NaN  34.0444 -118.2628  
2   1400 W  37TH                         ST          NaN  34.0210 -118.3002  
3  14000    RIVERSIDE                    DR          NaN  34.1576 -118.4387  
4                          200 E  AVENUE 28          NaN  34.0820 -118.2130  

[5 rows x 28 columns]

In [258]:
crime_df.shape

(1003448, 28)

The dataset has over one million rows, which means there will be plenty of data to investigate patterns in crime on different days, different times, the distribution of different kinds of crime etc. Before continuing to look at the data, I want to change the column names so it will be clearer what they mean. The source of the data on Kaggle has a description of each column which I can use to create more appropriate column titles.

In [260]:
crime_df.columns = ["Identifier", "Date_Reported", "Date_Occurred", "Time_Occurred", "Area_Code", "Area_Name", \
                    "District_Number", "Crime_Severity", "Crime_Type_Code", "Crime_Desc", "Crime_Method", \
                    "Victim_Age", "Victim_Gender", "Victim_Ethnicity", "Type_of_Premises_Code", \
                    "Type_of_Premises_Desc", "Weapon_Used", "Weapon_Desc", "Crime_Status_Code", 
                    "Crime_Status_Desc", "Crime_Code_1", "Crime_Code_2", "Crime_Code_3", "Crime_Code_4", \
                    "Crime_Location", "Cross_Street", "Latitude", "Longitude"]
crime_df.head()

Identifier           Date_Reported           Date_Occurred  Time_Occurred  \
0   190326475  03/01/2020 12:00:00 AM  03/01/2020 12:00:00 AM           2130   
1   200106753  02/09/2020 12:00:00 AM  02/08/2020 12:00:00 AM           1800   
2   200320258  11/11/2020 12:00:00 AM  11/04/2020 12:00:00 AM           1700   
3   200907217  05/10/2023 12:00:00 AM  03/10/2020 12:00:00 AM           2037   
4   200412582  09/09/2020 12:00:00 AM  09/09/2020 12:00:00 AM            630   

   Area_Code   Area_Name  District_Number  Crime_Severity  Crime_Type_Code  \
0          7    Wilshire              784               1              510   
1          1     Central              182               1              330   
2          3   Southwest              356               1              480   
3          9    Van Nuys              964               1              343   
4          4  Hollenbeck              413               1              510   

                                 Crime_Desc  ... Crime_Status_Code  \
0                          VEHICLE - STOLEN  ...                AA   
1                     BURGLARY FROM VEHICLE  ...                IC   
2                             BIKE - STOLEN  ...                IC   
3  SHOPLIFTING-GRAND THEFT ($950.01 & OVER)  ...                IC   
4                          VEHICLE - STOLEN  ...                IC   

   Crime_Status_Desc Crime_Code_1 Crime_Code_2  Crime_Code_3 Crime_Code_4  \
0       Adult Arrest        510.0        998.0           NaN          NaN   
1        Invest Cont        330.0        998.0           NaN          NaN   
2        Invest Cont        480.0          NaN           NaN          NaN   
3        Invest Cont        343.0          NaN           NaN          NaN   
4        Invest Cont        510.0          NaN           NaN          NaN   

                             Crime_Location Cross_Street Latitude Longitude  
0   1900 S  LONGWOOD                     AV          NaN  34.0375 -118.3506  
1   1000 S  FLOWER                       ST          NaN  34.0444 -118.2628  
2   1400 W  37TH                         ST          NaN  34.0210 -118.3002  
3  14000    RIVERSIDE                    DR          NaN  34.1576 -118.4387  
4                          200 E  AVENUE 28          NaN  34.0820 -118.2130  

[5 rows x 28 columns]

# Dealing with Missing Data

In [262]:
crime_df.isna().sum()

Identifier                     0
Date_Reported                  0
Date_Occurred                  0
Time_Occurred                  0
Area_Code                      0
Area_Name                      0
District_Number                0
Crime_Severity                 0
Crime_Type_Code                0
Crime_Desc                     0
Crime_Method              151567
Victim_Age                     0
Victim_Gender             144592
Victim_Ethnicity          144604
Type_of_Premises_Code         16
Type_of_Premises_Desc        588
Weapon_Used               676308
Weapon_Desc               676308
Crime_Status_Code              1
Crime_Status_Desc              0
Crime_Code_1                  11
Crime_Code_2              934330
Crime_Code_3             1001133
Crime_Code_4             1003384
Crime_Location                 0
Cross_Street              849335
Latitude                       0
Longitude                      0
dtype: int64

Overall there is not much data missing from the dataset. Crime_Method, Victim_Age and Victim_Gender all have around 150k values missing, which is not a huge problem considering there are over a million rows in total. I won't remove these rows because they will likely still contain useful information about geographic distribution of crime etc. I will remove rows with Type_of_Premises_Code missing because there are only 16 rows missing and this will clean up the data with minimal impact.

In [264]:
crime_df = crime_df[crime_df["Type_of_Premises_Code"].notna()]
crime_df["Type_of_Premises_Desc"].isna().sum()

572

There are still 572 missing values for Type_of_Premises_Desc. It is possible that I could get this information from Type_of_Premises_Code.

In [266]:
crime_df[crime_df["Type_of_Premises_Desc"].isna()]["Type_of_Premises_Code"].value_counts()

Type_of_Premises_Code
418.0    372
256.0    164
972.0     15
974.0     11
973.0      7
976.0      2
975.0      1
Name: count, dtype: int64

In [267]:
for code in [418.0, 256.0, 972.0, 974.0, 973.0, 976.0, 975.0]:
    print(crime_df.query("Type_of_Premises_Code == @code")["Type_of_Premises_Desc"].value_counts())

Series([], Name: count, dtype: int64)
Series([], Name: count, dtype: int64)
Series([], Name: count, dtype: int64)
Series([], Name: count, dtype: int64)
Series([], Name: count, dtype: int64)
Series([], Name: count, dtype: int64)
Series([], Name: count, dtype: int64)


Unfortunately no information can be gained from Type_of_Premises_Code. So I will remove all rows with null values for Type_of_Premises_Desc.

In [269]:
crime_df = crime_df[crime_df["Type_of_Premises_Desc"].notna()]

Weapon_Used and Weapon_Desc both have almost 700k values missing. This represents a large proportion of the total dataset size, but 300k values is still plenty for analysis. So I will leave these columns for now. The Crime_Code columns allow for more then one crime code if there were multiple offenses in one crime. Having null values for the last 3 is expected, since most crimes only involve one offense, but the 11 missing values for Crime_Code_1 should be investigated.

In [271]:
crime_df[crime_df["Crime_Code_1"].isna()]

Identifier           Date_Reported           Date_Occurred  \
5790     200112035  05/21/2020 12:00:00 AM  05/21/2020 12:00:00 AM   
47724    200116522  08/17/2020 12:00:00 AM  08/17/2020 12:00:00 AM   
236992   211016055  12/26/2021 12:00:00 AM  12/25/2021 12:00:00 AM   
350446   210118616  10/13/2021 12:00:00 AM  10/13/2021 12:00:00 AM   
432346   221701255  12/08/2022 12:00:00 AM  12/08/2022 12:00:00 AM   
495761   220509912  06/06/2022 12:00:00 AM  06/06/2022 12:00:00 AM   
508440   220805565  02/09/2022 12:00:00 AM  02/09/2022 12:00:00 AM   
694430   230114704  06/22/2023 12:00:00 AM  06/22/2023 12:00:00 AM   
705361   230123939  11/13/2023 12:00:00 AM  11/12/2023 12:00:00 AM   
801472   230106125  02/01/2023 12:00:00 AM  02/01/2023 12:00:00 AM   
873795   230114706  06/22/2023 12:00:00 AM  06/22/2023 12:00:00 AM   

        Time_Occurred  Area_Code    Area_Name  District_Number  \
5790              525          1      Central              163   
47724             545          1      Central              176   
236992           2000         10  West Valley             1023   
350446           1840          1      Central              162   
432346           2138         17   Devonshire             1799   
495761           2000          5       Harbor              526   
508440            245          8      West LA              842   
694430           1801          1      Central              162   
705361           2300          1      Central              162   
801472           1855          1      Central              129   
873795           1550          1      Central              155   

        Crime_Severity  Crime_Type_Code  \
5790                 2              740   
47724                2              745   
236992               1              520   
350446               1              442   
432346               2              890   
495761               2              888   
508440               2              888   
694430               1              442   
705361               1              330   
801472               1              330   
873795               2              623   

                                               Crime_Desc  ...  \
5790    VANDALISM - FELONY ($400 & OVER, ALL CHURCH VA...  ...   
47724            VANDALISM - MISDEAMEANOR ($399 OR UNDER)  ...   
236992                           VEHICLE - ATTEMPT STOLEN  ...   
350446           SHOPLIFTING - PETTY THEFT ($950 & UNDER)  ...   
432346                                   FAILURE TO YIELD  ...   
495761                                        TRESPASSING  ...   
508440                                        TRESPASSING  ...   
694430           SHOPLIFTING - PETTY THEFT ($950 & UNDER)  ...   
705361                              BURGLARY FROM VEHICLE  ...   
801472                              BURGLARY FROM VEHICLE  ...   
873795                            BATTERY POLICE (SIMPLE)  ...   

       Crime_Status_Code  Crime_Status_Desc Crime_Code_1 Crime_Code_2  \
5790                  IC        Invest Cont          NaN        740.0   
47724                 IC        Invest Cont          NaN        745.0   
236992                IC        Invest Cont          NaN        520.0   
350446                IC        Invest Cont          NaN        442.0   
432346                IC        Invest Cont          NaN        890.0   
495761                IC        Invest Cont          NaN        888.0   
508440                IC        Invest Cont          NaN        888.0   
694430                IC        Invest Cont          NaN        442.0   
705361                IC        Invest Cont          NaN        330.0   
801472                IC        Invest Cont          NaN        330.0   
873795                IC        Invest Cont          NaN        623.0   

        Crime_Code_3 Crime_Code_4                            Crime_Location  \
5790             NaN          NaN                                       9TH   
47724            NaN         

Each of these crimes involved just one offense which was put in Crime_Code_2. This was probably a mistake made by whoever was responsible for entering the information into the system. I will move the values from Crime_Code_2 into Crime_Code_1.

In [273]:
idx = crime_df["Crime_Code_1"].isna()
crime_df.loc[idx] = crime_df.loc[idx].rename(columns={"Crime_Code_1": "Crime_Code_2", "Crime_Code_2": "Crime_Code_1"})

Finally, Cross_Street has around 850k values missing. This is a considerable amount of missing data but I will leave it for now. It isn't very useful so will probably be removed along with other columns I don't need.

In [275]:
crime_df.shape

(1002860, 28)

The total number of entries in the dataset has been reduced from 1,003,448 to 1,002,860. Only around 600 rows were removed, which means the original dataset was already very good. Now I will look at the columns and see if there are any I want to remove straight away.

# Removing Unncessary Columns

In [278]:
crime_df.columns

Index(['Identifier', 'Date_Reported', 'Date_Occurred', 'Time_Occurred',
       'Area_Code', 'Area_Name', 'District_Number', 'Crime_Severity',
       'Crime_Type_Code', 'Crime_Desc', 'Crime_Method', 'Victim_Age',
       'Victim_Gender', 'Victim_Ethnicity', 'Type_of_Premises_Code',
       'Type_of_Premises_Desc', 'Weapon_Used', 'Weapon_Desc',
       'Crime_Status_Code', 'Crime_Status_Desc', 'Crime_Code_1',
       'Crime_Code_2', 'Crime_Code_3', 'Crime_Code_4', 'Crime_Location',
       'Cross_Street', 'Latitude', 'Longitude'],
      dtype='object')

The Identifier column is probably useful for the organisation that produced the data but does not help me. It will be easier to use the id that Pandas automatically creates for the data.

In [280]:
crime_df = crime_df.drop("Identifier", axis=1)

### Geographical Columns

Date_Reported, Date_Occurred and Time_Occurred are all useful information. In terms of geographical information, there are these columns: Area_Code, Area_Name, District_Number, Crime_Location, Cross_Street, Latitude, Longitude. Latitude and Longitude will be most important for mapping the data in Tableau. I want to see what values Area_Name takes on.

In [283]:
crime_df["Area_Name"].value_counts()

Area_Name
Central        69525
77th Street    61637
Pacific        59298
Southwest      57313
Hollywood      52331
N Hollywood    50986
Olympic        49984
Southeast      49886
Newton         49009
Wilshire       48116
Rampart        46767
West LA        45663
Northeast      42849
Van Nuys       42798
West Valley    42073
Devonshire     41657
Harbor         41292
Topanga        41286
Mission        40283
Hollenbeck     37025
Foothill       33082
Name: count, dtype: int64

There aren't too many area names and they all have a decent amount of data, so it could be useful for analysing which neighbourhoods have higher crime rates, for example. Area_Code is redundant so I will remove that column.

In [285]:
crime_df = crime_df.drop("Area_Code", axis=1)

District_Number represents the district of the department that responded to the crime. This could be useful for analysing which departments have to deal with more crimes and if this changes over time, and therefore which districts should receieve more funding etc.

In [287]:
crime_df["Crime_Location"].value_counts()

Crime_Location
800 N  ALAMEDA                      ST      2597
700 S  FIGUEROA                     ST      1705
100    THE GROVE                    DR      1668
10200    SANTA MONICA                 BL    1652
6TH                          ST             1585
                                            ... 
9800    OMELVENY                     ST        1
02400    MARIONDALE                   AV       1
15400    COPPER                       ST       1
4200    TOSCA                        RD        1
12200    ART                          ST       1
Name: count, Length: 66544, dtype: int64

In [288]:
crime_df["Cross_Street"].value_counts()

Cross_Street
BROADWAY                           2486
FIGUEROA                     ST    1933
FIGUEROA                           1646
VERMONT                      AV    1633
WESTERN                      AV    1489
                                   ... 
GOULD                        AV       1
MARTIN LURTH                          1
INGRAHAM                     AV       1
ALVERN                       AV       1
LUCERNE                      ST       1
Name: count, Length: 10404, dtype: int64

Crime_Location and Cross_Street represent very detailed geographical location and have lots of values which only occur once or a few times. This makes it hard to find use for these columns, and Longitude and Latitude already provide detailed geographic information. So I will remove these two columns.

In [290]:
crime_df = crime_df.drop(["Crime_Location", "Cross_Street"], axis=1)

### Columns Describing the Nature of the Crime

There are some more columns that describe the nature of the crime: Crime_Severity, Crime_Type_Code, Crime_Desc, Crime_Method, Weapon_Used, Weapon_Desc, and the Crime_Code columns. Firstly, Crime_Type_Code and Weapon_Used are redundant and can be removed.

In [293]:
crime_df = crime_df.drop(["Crime_Type_Code", "Weapon_Used"], axis=1)

Now I want to see what kinds of values the other columns take on.

In [295]:
crime_df["Crime_Severity"].value_counts()

Crime_Severity
1    601027
2    401833
Name: count, dtype: int64

Going back to Kaggle, I can see that 1 represents serious crimes and 2 represents less serious crimes. This could be useful information.

In [297]:
crime_df["Crime_Desc"].value_counts()

Crime_Desc
VEHICLE - STOLEN                                           115182
BATTERY - SIMPLE ASSAULT                                    74798
BURGLARY FROM VEHICLE                                       63259
THEFT OF IDENTITY                                           62407
VANDALISM - FELONY ($400 & OVER, ALL CHURCH VANDALISMS)     60950
                                                            ...  
FIREARMS EMERGENCY PROTECTIVE ORDER (FIREARMS EPO)              5
FIREARMS RESTRAINING ORDER (FIREARMS RO)                        4
DISHONEST EMPLOYEE ATTEMPTED THEFT                              4
TRAIN WRECKING                                                  1
DRUNK ROLL - ATTEMPT                                            1
Name: count, Length: 140, dtype: int64

Again this is very useful information to analyse the types of crime that are committed. For values that only have a few entries, it might be worth changing their value to "Other" since there is too little information about them, but that is a job for later.

In [299]:
crime_df["Crime_Method"].value_counts()

Crime_Method
0344                                                 42176
0329                                                 21581
1822 0344                                            14169
0325                                                 10300
1501                                                  9809
                                                     ...  
0449 0337 1822                                           1
1822 1906 0701                                           1
2000 0400 0416 0913 1814 0446 0417 0381 1414             1
1822 0216 0329 0334 0352 0421 1601 0382 1402 0344        1
2047 0400 0429 0444 1202 1822                            1
Name: count, Length: 310294, dtype: int64

Looking back at the dataset description on Kaggle, these are the modus operandi codes that show in more detail how the crime occured. For example, the 0329 that appears quite a lot represents "vandalised". A long list of codes could be used to analyse the exact way the crime was carried out, but this is a different kind of analysis to the one I want to do. So I will remove this column from the dataset.

In [301]:
crime_df = crime_df.drop("Crime_Method", axis=1)
crime_df["Weapon_Desc"].value_counts()

Weapon_Desc
STRONG-ARM (HANDS, FIST, FEET OR BODILY FORCE)    174672
UNKNOWN WEAPON/OTHER WEAPON                        36308
VERBAL THREAT                                      23834
HAND GUN                                           20135
SEMI-AUTOMATIC PISTOL                               7242
                                                   ...  
MAC-11 SEMIAUTOMATIC ASSAULT WEAPON                    3
MAC-10 SEMIAUTOMATIC ASSAULT WEAPON                    2
HECKLER & KOCH 91 SEMIAUTOMATIC ASSAULT RIFLE          2
M-14 SEMIAUTOMATIC ASSAULT RIFLE                       2
M1-1 SEMIAUTOMATIC ASSAULT RIFLE                       1
Name: count, Length: 79, dtype: int64

The Weapon_Desc is a useful feature to analyse, similar to Crime_Desc. The final columns related to the actual crime to consider are the Crime_Code columns. According to Kaggle, the Crime_Code_1 column is actually the same as the Crime_Type_Code column that I removed earlier. The other Crime_Code columns are there to record additional offenses in the same crime. So Crime_Code_1 can definitely be removed since it is redundant. I want to check how many values there are for the other columns.

In [303]:
print("Crime_Code_2:", crime_df["Crime_Code_2"].notna().sum(), "non-null values")
print("Crime_Code_3:", crime_df["Crime_Code_3"].notna().sum(), "non-null values")
print("Crime_Code_4:", crime_df["Crime_Code_4"].notna().sum(), "non-null values")

Crime_Code_2: 68960 non-null values
Crime_Code_3: 2315 non-null values
Crime_Code_4: 64 non-null values


There definitely isn't enough data for Crime_Code_3 and Crime_Code_4. Crime_Code_2 potentially has enough data for analysis but I can't see how it would fit in with my current plan for this project. So all of the Crime_Code columns should be removed.

In [305]:
crime_df = crime_df.drop(["Crime_Code_1", "Crime_Code_2", "Crime_Code_3", "Crime_Code_4"], axis=1)
crime_df.columns

Index(['Date_Reported', 'Date_Occurred', 'Time_Occurred', 'Area_Name',
       'District_Number', 'Crime_Severity', 'Crime_Desc', 'Victim_Age',
       'Victim_Gender', 'Victim_Ethnicity', 'Type_of_Premises_Code',
       'Type_of_Premises_Desc', 'Weapon_Desc', 'Crime_Status_Code',
       'Crime_Status_Desc', 'Latitude', 'Longitude'],
      dtype='object')

### Other Columns

The columns that remain to be checked are Victim_Age, Victim_Gender, Victim_Ethnicity, Type_of_Premises_Code, Type_of_Premises_Desc, Crime_Status_Code and Crime_Status_Desc. Type_of_Premises_Code and Crime_Status_Code are redundant so can be removed straight away.

In [308]:
crime_df = crime_df.drop(["Type_of_Premises_Code", "Crime_Status_Code"], axis=1)

Victim_Age, Victim_Gender and Victim_Ethnicity will definitely be useful for analysis. So the last thing to do is check what values are taken on by Type_of_Premises_Desc and Crime_Status_Desc.

In [310]:
crime_df["Type_of_Premises_Desc"].value_counts()

Type_of_Premises_Desc
STREET                                          261032
SINGLE FAMILY DWELLING                          163477
MULTI-UNIT DWELLING (APARTMENT, DUPLEX, ETC)    118849
PARKING LOT                                      69049
OTHER BUSINESS                                   47565
                                                 ...  
MTA - SILVER LINE - LAC/USC MEDICAL CENTER           2
DEPT OF DEFENSE FACILITY                             2
HORSE RACING/SANTA ANITA PARK*                       2
MTA - SILVER LINE - DOWNTOWN STREET STOPS            2
TRAM/STREETCAR(BOXLIKE WAG ON RAILS)*                1
Name: count, Length: 306, dtype: int64

In [311]:
crime_df["Crime_Status_Desc"].value_counts()

Crime_Status_Desc
Invest Cont     802961
Adult Other     108504
Adult Arrest     86300
Juv Arrest        3239
Juv Other         1850
UNK                  6
Name: count, dtype: int64

Both columns have a fairly short range of values and enough data for categorical analysis, so I will keep both of them. I notice that Crime_Status_Desc has UNK as a value which will need to be dealt with, but that's a job for a different section. All unnecessary columns have been removed and I can have a look at what the dataset looks like now.

In [313]:
crime_df = crime_df.reset_index(drop=True)
crime_df.head()

Date_Reported           Date_Occurred  Time_Occurred   Area_Name  \
0  03/01/2020 12:00:00 AM  03/01/2020 12:00:00 AM           2130    Wilshire   
1  02/09/2020 12:00:00 AM  02/08/2020 12:00:00 AM           1800     Central   
2  11/11/2020 12:00:00 AM  11/04/2020 12:00:00 AM           1700   Southwest   
3  05/10/2023 12:00:00 AM  03/10/2020 12:00:00 AM           2037    Van Nuys   
4  09/09/2020 12:00:00 AM  09/09/2020 12:00:00 AM            630  Hollenbeck   

   District_Number  Crime_Severity                                Crime_Desc  \
0              784               1                          VEHICLE - STOLEN   
1              182               1                     BURGLARY FROM VEHICLE   
2              356               1                             BIKE - STOLEN   
3              964               1  SHOPLIFTING-GRAND THEFT ($950.01 & OVER)   
4              413               1                          VEHICLE - STOLEN   

   Victim_Age Victim_Gender Victim_Ethnicity  \
0           0             M                O   
1          47             M                O   
2          19             X                X   
3          19             M                O   
4           0           NaN              NaN   

                          Type_of_Premises_Desc Weapon_Desc Crime_Status_Desc  \
0                                        STREET         NaN      Adult Arrest   
1             BUS STOP/LAYOVER (ALSO QUERY 124)         NaN       Invest Cont   
2  MULTI-UNIT DWELLING (APARTMENT, DUPLEX, ETC)         NaN       Invest Cont   
3                                CLOTHING STORE         NaN       Invest Cont   
4                                        STREET         NaN       Invest Cont   

   Latitude  Longitude  
0   34.0375  -118.3506  
1   34.0444  -118.2628  
2   34.0210  -118.3002  
3   34.1576  -118.4387  
4   34.0820  -118.2130

In [314]:
crime_df.shape

(1002860, 15)

# Cleaning Columns

Now I want to look at each column individually and make sure the data is represented in a consistent way throughout the dataset. I want to turn the data into a suitable datatype and ensure the data is immediately understandable.

### Time Columns

The Date_Reported and Date_Occured columns include the date and time, although the time seems to be 12AM throughout the whole dataset.

In [319]:
crime_df["Date_Reported"].dtype

dtype('O')

The data is currently stored as an object, not a datetime, so I will start by removing the time component in each row.

In [321]:
crime_df["Date_Reported"] = crime_df["Date_Reported"].str.replace("12:00:00 AM", "")
crime_df["Date_Occurred"] = crime_df["Date_Occurred"].str.replace("12:00:00 AM", "")

In [322]:
crime_df["Date_Reported"] = pd.to_datetime(crime_df["Date_Reported"])
crime_df["Date_Occurred"] = pd.to_datetime(crime_df["Date_Occurred"])

In [323]:
crime_df["Time_Occurred"].value_counts()

Time_Occurred
1200    35160
1800    26530
1700    25151
2000    24741
1900    23029
        ...  
741        29
534        26
2332       25
559        23
531        19
Name: count, Length: 1439, dtype: int64

There are definitely some problems with the Time_Occurred column. It should be in 24-hour format but some values are only three digits long. I'll see how many values are four digits long.

In [325]:
crime_df["Time_Occurred"].astype("string").str.len().value_counts()

Time_Occurred
4    726636
3    235843
1     21488
2     18893
Name: count, dtype: Int64

When the time is three digits long, there is no way to know if it is AM or PM, or indeed if the data has been entered correctly. The only solution I can think of is to set invalid times as null and accept that there is no time component for around a quarter of the data.

In [327]:
time_lengths = crime_df["Time_Occurred"].astype("string").str.len()
crime_df["Time_Occurred"] = crime_df["Time_Occurred"].mask(time_lengths.isin([1,2,3]), np.nan)

In [328]:
crime_df["Time_Occurred"] = pd.to_datetime(crime_df["Time_Occurred"], format="%H%M").dt.time

Now the time is in a standard format and invalid times have been changed to NaT.

### Area Name and District Number

In [331]:
crime_df["Area_Name"].value_counts()

Area_Name
Central        69525
77th Street    61637
Pacific        59298
Southwest      57313
Hollywood      52331
N Hollywood    50986
Olympic        49984
Southeast      49886
Newton         49009
Wilshire       48116
Rampart        46767
West LA        45663
Northeast      42849
Van Nuys       42798
West Valley    42073
Devonshire     41657
Harbor         41292
Topanga        41286
Mission        40283
Hollenbeck     37025
Foothill       33082
Name: count, dtype: int64

Area_Name looks fine! Onto the next column.

In [333]:
crime_df["District_Number"].value_counts()

District_Number
162     5390
1494    5317
645     5014
182     4884
646     4416
        ... 
1784       1
225        1
1053       1
1087       1
902        1
Name: count, Length: 1210, dtype: int64

In [334]:
crime_df.dtypes

Date_Reported            datetime64[ns]
Date_Occurred            datetime64[ns]
Time_Occurred                    object
Area_Name                        object
District_Number                   int64
Crime_Severity                    int64
Crime_Desc                       object
Victim_Age                        int64
Victim_Gender                    object
Victim_Ethnicity                 object
Type_of_Premises_Desc            object
Weapon_Desc                      object
Crime_Status_Desc                object
Latitude                        float64
Longitude                       float64
dtype: object

District_Number has no obvious problems and since Pandas correctly interpreted it as int64, everything looks fine.

### Crime Severity and Description

In [337]:
crime_df["Crime_Severity"].value_counts()

Crime_Severity
1    601027
2    401833
Name: count, dtype: int64

There are no issues with Crime_Severity but I want to change the values to text to make its meaning clearer.

In [339]:
crime_df["Crime_Severity"] = crime_df["Crime_Severity"].map({1: "Severe", 2: "Not Severe"})

In [340]:
crime_df["Crime_Desc"].value_counts()

Crime_Desc
VEHICLE - STOLEN                                           115182
BATTERY - SIMPLE ASSAULT                                    74798
BURGLARY FROM VEHICLE                                       63259
THEFT OF IDENTITY                                           62407
VANDALISM - FELONY ($400 & OVER, ALL CHURCH VANDALISMS)     60950
                                                            ...  
FIREARMS EMERGENCY PROTECTIVE ORDER (FIREARMS EPO)              5
FIREARMS RESTRAINING ORDER (FIREARMS RO)                        4
DISHONEST EMPLOYEE ATTEMPTED THEFT                              4
TRAIN WRECKING                                                  1
DRUNK ROLL - ATTEMPT                                            1
Name: count, Length: 140, dtype: int64

There are lots of values in Crime_Desc that don't have enough data for useful analysis. So I'll combine all values that have less than 1000 entries into one "Other" value.

In [342]:
value_counts = crime_df["Crime_Desc"].value_counts()
rare_values = value_counts[value_counts < 1000].index
crime_df["Crime_Desc"] = crime_df["Crime_Desc"].replace(rare_values, "Other")

In [343]:
crime_df["Crime_Desc"].value_counts()

Crime_Desc
VEHICLE - STOLEN                                            115182
BATTERY - SIMPLE ASSAULT                                     74798
BURGLARY FROM VEHICLE                                        63259
THEFT OF IDENTITY                                            62407
VANDALISM - FELONY ($400 & OVER, ALL CHURCH VANDALISMS)      60950
BURGLARY                                                     57590
ASSAULT WITH DEADLY WEAPON, AGGRAVATED ASSAULT               53496
THEFT PLAIN - PETTY ($950 & UNDER)                           53465
INTIMATE PARTNER - SIMPLE ASSAULT                            46709
THEFT FROM MOTOR VEHICLE - PETTY ($950 & UNDER)              41263
THEFT FROM MOTOR VEHICLE - GRAND ($950.01 AND OVER)          36810
THEFT-GRAND ($950.01 & OVER)EXCPT,GUNS,FOWL,LIVESTK,PROD     34999
ROBBERY                                                      32244
SHOPLIFTING - PETTY THEFT ($950 & UNDER)                     30728
VANDALISM - MISDEAMEANOR ($399 OR UNDER)           

Now that there's a sensible range of values, I want to rename them to make the crime descriptions easier to understand.

In [345]:
crime_df["Crime_Desc"] = crime_df["Crime_Desc"].map({"VEHICLE - STOLEN": "Stolen Vehicle", 
                                                     "BATTERY - SIMPLE ASSAULT": "Simple Assault",
                                                     "BURGLARY FROM VEHICLE": "Burglary from Vehicle",
                                                     "THEFT OF IDENTITY": "Identity Theft",
                                                     "VANDALISM - FELONY ($400 & OVER, ALL CHURCH VANDALISMS)": "Felony Vandalism",
                                                     "ASSAULT WITH DEADLY WEAPON, AGGRAVATED ASSAULT": "Assault with Deadly Weapon",
                                                     "THEFT PLAIN - PETTY ($950 & UNDER)": "Petty Plain Theft",
                                                     "INTIMATE PARTNER - SIMPLE ASSAULT": "Simple Assault on Partner",
                                                     "THEFT FROM MOTOR VEHICLE - PETTY ($950 & UNDER)": "Petty Theft from Vehicle",
                                                     "THEFT FROM MOTOR VEHICLE - GRAND ($950.01 AND OVER)": "Grand Theft from Vehicle",
                                                     "THEFT-GRAND ($950.01 & OVER)EXCPT,GUNS,FOWL,LIVESTK,PROD": "Grand Theft",
                                                     "SHOPLIFTING - PETTY THEFT ($950 & UNDER)": "Petty Shoplifting",
                                                     "VANDALISM - MISDEAMEANOR ($399 OR UNDER)": "Misdeameanor Vandalism",
                                                     "CRIMINAL THREATS - NO WEAPON DISPLAYED": "Criminal Threats with no Weapon",
                                                     "INTIMATE PARTNER - AGGRAVATED ASSAULT": "Aggravated Assault on Partner",
                                                     "LETTERS, LEWD  -  TELEPHONE CALLS, LEWD": "Harassing Communication",
                                                     "BIKE - STOLEN": "Stolen Bike",
                                                     "OTHER MISCELLANEOUS CRIME": "Other",
                                                     "BUNCO, GRAND THEFT": "Bunco Grand Theft",
                                                     "SHOPLIFTING-GRAND THEFT ($950.01 & OVER)": "Grand Shoplifting",
                                                     "EMBEZZLEMENT, GRAND THEFT ($950.01 & OVER)": "Grand Embezzelment",
                                                     "THEFT, PERSON": "Theft from Person",
                                                     "BURGLARY, ATTEMPTED": "Attempted Burglary",
                                                     "VEHICLE - ATTEMPT STOLEN": "Attempted Stolen Vehicle",
                                                     "DISCHARGE FIREARMS/SHOTS FIRED": "Shots Fired",
                                                     "RAPE, FORCIBLE": "Forcible Rape",
                                                     "CHILD ABUSE (PHYSICAL) - SIMPLE ASSAULT": "Child Abuse",
                                                     "DOCUMENT FORGERY / STOLEN FELONY": "Document Forgery",
                                                     "VEHICLE, STOLEN - OTHER (MOTORIZED SCOOTERS, BIKES, ETC)": "Stolen Vehicle",
                                                     "BATTERY POLICE (SIMPLE)": "Battery on Police",
                                                     "BUNCO, PETTY THEFT": "Bunco Petty Theft",
                                                     "CRM AGNST CHLD (13 OR UNDER) (14-15 & SUSP 10 YRS OLDER)": "Criminal Offense against Child",
                                                     "SEXUAL PENETRATION W/FOREIGN OBJECT": "Sexual Penetration with Object",
                                                     "CHILD NEGLECT (SEE 300 W.I.C.)": "Child Neglect",
                                                     "SEX,UNLAWFUL(INC MUTUAL CONSENT, PENETRATION W/ FRGN OBJ": "Unlawful Sex",
                                                     "CHILD ANNOYING (17YRS & UNDER)": "Harassing a Child"}).fillna(crime_df["Crime_Desc"])
crime_df["Crime_Desc"] = crime_df["Crime_Desc"].str.title()

### Vitctim Age, Gender and Ethnicity

The next column to check is Victim_Age, and I can already see from the head of the dataset that there are some values of 0. 

In [348]:
crime_df["Victim_Age"].value_counts().sort_index()

Victim_Age
-4           3
-3           5
-2          28
-1          99
 0      268562
         ...  
 96         95
 97         72
 98         71
 99        354
 120         1
Name: count, Length: 104, dtype: int64

It looks like there's no official way for them to indicate that the age is not known, so the value depends on who is entering the data. Any negative ages are clearly invalid and a 0 probably indicates the age was not available. The number of entries for ages in the 90s decreases steadily but suddenly increases again at 99, which could suggest some people chose 99 to indicate lack of data. 120 is also an outlier and probably invalid. So, I will change the following values to null: -4, -3, -2, -1, 0, 99, 120.

In [350]:
crime_df["Victim_Age"] = crime_df["Victim_Age"].apply(lambda x: np.nan if x in [-4,-3,-2,-1,0,99,120] else x)

In [351]:
crime_df["Victim_Gender"].value_counts()

Victim_Gender
M    402930
F    357960
X     97279
H       114
-         1
Name: count, dtype: int64

The data source on Kaggle indicates that M means Male, F means Female and X means Unknown. The meaning of H and - are not specified so I will assume that these genders are also unknown. So I will change the values of X, H and - to null.

In [353]:
crime_df["Victim_Gender"] = crime_df["Victim_Gender"].apply(lambda x: np.nan if x in ["X","H","-"] else x)
crime_df["Victim_Gender"] = crime_df["Victim_Gender"].map({"M": "Male", "F": "Female"})

In [354]:
crime_df["Victim_Ethnicity"].unique()

array(['O', 'X', nan, 'W', 'H', 'B', 'A', 'K', 'C', 'F', 'Z', 'V', 'I',
       'P', 'J', 'G', 'U', 'D', 'S', 'L', '-'], dtype=object)

The ethnicities represented by these letters is not specified on Kaggle. By looking at other crime statistics, it is possible to deduce that H=Hispanic, W=White, B=African American, A=Asian. I will assume that O=Other and group the other letters together as "Unknown".

In [356]:
crime_df["Victim_Ethnicity"] = crime_df["Victim_Ethnicity"].map({"O": "Other",
                                                                 "H": "Hispanic",
                                                                 "W": "White",
                                                                 "B": "African American",
                                                                 "A": "Asian"}).fillna("Unknown")

### Type of Premises

In [358]:
crime_df["Type_of_Premises_Desc"].value_counts()

Type_of_Premises_Desc
STREET                                          261032
SINGLE FAMILY DWELLING                          163477
MULTI-UNIT DWELLING (APARTMENT, DUPLEX, ETC)    118849
PARKING LOT                                      69049
OTHER BUSINESS                                   47565
                                                 ...  
MTA - SILVER LINE - LAC/USC MEDICAL CENTER           2
DEPT OF DEFENSE FACILITY                             2
HORSE RACING/SANTA ANITA PARK*                       2
MTA - SILVER LINE - DOWNTOWN STREET STOPS            2
TRAM/STREETCAR(BOXLIKE WAG ON RAILS)*                1
Name: count, Length: 306, dtype: int64

Similarly to with Crime_Desc, I will group all values with less than 1000 occurences under "Other".

In [360]:
value_counts = crime_df["Type_of_Premises_Desc"].value_counts()
rare_values = value_counts[value_counts < 1000].index
crime_df["Type_of_Premises_Desc"] = crime_df["Type_of_Premises_Desc"].replace(rare_values, "Other")

In [361]:
crime_df["Type_of_Premises_Desc"].unique()

array(['STREET', 'Other', 'MULTI-UNIT DWELLING (APARTMENT, DUPLEX, ETC)',
       'CLOTHING STORE', 'PUBLIC STORAGE', 'OTHER BUSINESS',
       'PARKING LOT', 'ALLEY', 'GAS STATION', 'SINGLE FAMILY DWELLING',
       'CONDOMINIUM/TOWNHOUSE', 'RESTAURANT/FAST FOOD', 'SIDEWALK',
       'NURSING/CONVALESCENT/RETIREMENT HOME', 'MARKET',
       'GOVERNMENT FACILITY (FEDERAL,STATE, COUNTY & CITY)', 'DRIVEWAY',
       'MINI-MART', 'YARD (RESIDENTIAL/BUSINESS)',
       'VEHICLE, PASSENGER/TRUCK', 'GARAGE/CARPORT',
       'PARKING UNDERGROUND/BUILDING', 'PORCH, RESIDENTIAL', 'DRUG STORE',
       'MTA BUS', 'OTHER PREMISE', 'BUS STOP', 'POLICE FACILITY',
       'MISSIONS/SHELTERS',
       'CHURCH/CHAPEL (CHANGED 03-03 FROM CHURCH/TEMPLE)',
       'DISCOUNT STORE (99 CENT,DOLLAR,ETC.', 'OFFICE BUILDING/OFFICE',
       'OTHER STORE', 'HOTEL', 'OTHER RESIDENCE', 'OTHER/OUTSIDE',
       'JUNIOR HIGH SCHOOL', 'NIGHT CLUB (OPEN EVENINGS ONLY)',
       "COFFEE SHOP (STARBUCKS, COFFEE BEAN, PEET'S, ETC.)",

In [362]:
crime_df["Type_of_Premises_Desc"] = (crime_df["Type_of_Premises_Desc"]
                                     .map({"MULTI-UNIT DWELLING (APARTMENT, DUPLEX, ETC)": "Multi-Unit Dwelling",
                                           "GOVERNMENT FACILITY (FEDERAL,STATE, COUNTY & CITY)": "Government Facility",
                                           "YARD (RESIDENTIAL/BUSINESS)": "Yard",
                                           "VEHICLE, PASSENGER/TRUCK": "Vehicle",
                                           "PORCH, RESIDENTIAL": "Residential Porch",
                                           "OTHER PREMISE": "Other",
                                           "CHURCH/CHAPEL (CHANGED 03-03 FROM CHURCH/TEMPLE)": "Church/Chapel",
                                           "DISCOUNT STORE (99 CENT,DOLLAR,ETC.": "Discount Store",
                                           "OFFICE BUILDING/OFFICE": "Office Building",
                                           "OTHER/OUTSIDE": "Other",
                                           "NIGHT CLUB (OPEN EVENINGS ONLY)": "Night Club",
                                           "COFFEE SHOP (STARBUCKS, COFFEE BEAN, PEET'S, ETC.)": "Coffee Shop",
                                           "LA UNION STATION (NOT LINE SPECIFIC)": "Union Station",
                                           "DIY CENTER (LOWE'S,HOME DEPOT,OSH,CONTRACTORS WAREHOUSE)": "DIY Center",
                                           "MTA - RED LINE - WESTLAKE/MACARTHUR PARK": "WESTLAKE/MACARTHUR PARK",
                                           "COLLEGE/JUNIOR COLLEGE/UNIVERSITY": "College/University",
                                           "SHOPPING MALL (COMMON AREA)": "Shopping Mall",
                                           "ABANDONED BUILDING ABANDONED HOUSE": "Abandoned Building",
                                           "AUTOMATED TELLER MACHINE (ATM)": "ATM",
                                           "BAR/COCKTAIL/NIGHTCLUB": "Bar",
                                           "TRANSPORTATION FACILITY (AIRPORT)": "Airport",
                                           "MTA - RED LINE - NORTH HOLLYWOOD": "North Hollywood"})).fillna(crime_df["Type_of_Premises_Desc"])
crime_df["Type_of_Premises_Desc"] = crime_df["Type_of_Premises_Desc"].str.title()

### Weapon Description

In [364]:
crime_df["Weapon_Desc"].value_counts()

Weapon_Desc
STRONG-ARM (HANDS, FIST, FEET OR BODILY FORCE)    174672
UNKNOWN WEAPON/OTHER WEAPON                        36308
VERBAL THREAT                                      23834
HAND GUN                                           20135
SEMI-AUTOMATIC PISTOL                               7242
                                                   ...  
MAC-11 SEMIAUTOMATIC ASSAULT WEAPON                    3
MAC-10 SEMIAUTOMATIC ASSAULT WEAPON                    2
HECKLER & KOCH 91 SEMIAUTOMATIC ASSAULT RIFLE          2
M-14 SEMIAUTOMATIC ASSAULT RIFLE                       2
M1-1 SEMIAUTOMATIC ASSAULT RIFLE                       1
Name: count, Length: 79, dtype: int64

Again, I want to group together those with less than 1000 entries and then rename the values appropriately...

In [366]:
value_counts = crime_df["Weapon_Desc"].value_counts()
rare_values = value_counts[value_counts < 1000].index
crime_df["Weapon_Desc"] = crime_df["Weapon_Desc"].replace(rare_values, "Other")

In [367]:
crime_df["Weapon_Desc"].value_counts()

Weapon_Desc
STRONG-ARM (HANDS, FIST, FEET OR BODILY FORCE)    174672
UNKNOWN WEAPON/OTHER WEAPON                        36308
VERBAL THREAT                                      23834
HAND GUN                                           20135
Other                                              12460
SEMI-AUTOMATIC PISTOL                               7242
KNIFE WITH BLADE 6INCHES OR LESS                    6833
UNKNOWN FIREARM                                     6576
OTHER KNIFE                                         5877
MACE/PEPPER SPRAY                                   3727
VEHICLE                                             3258
ROCK/THROWN OBJECT                                  2738
PIPE/METAL PIPE                                     2468
BOTTLE                                              2414
FOLDING KNIFE                                       2266
STICK                                               2248
CLUB/BAT                                            2088
KITCHEN KNIFE      

In [368]:
crime_df["Weapon_Desc"] = (crime_df["Weapon_Desc"]
                           .map({"STRONG-ARM (HANDS, FIST, FEET OR BODILY FORCE)": "Body",
                                 "UNKNOWN WEAPON/OTHER WEAPON": "Other",
                                 "KNIFE WITH BLADE 6INCHES OR LESS": "Knife",
                                 "KNIFE WITH BLADE OVER 6 INCHES IN LENGTH": "Knife",
                                 "OTHER KNIFE": "Knife"})).fillna(crime_df["Weapon_Desc"])
crime_df["Weapon_Desc"] = crime_df["Weapon_Desc"].str.title()

### Crime Status

In [370]:
crime_df["Crime_Status_Desc"].value_counts()

Crime_Status_Desc
Invest Cont     802961
Adult Other     108504
Adult Arrest     86300
Juv Arrest        3239
Juv Other         1850
UNK                  6
Name: count, dtype: int64

For Crime_Status_Desc, since there are so few rows with unknown value, I want to actually remove these rows entirely. Then, I will rename the remaining values so that they are easier to understand just like before.

In [372]:
crime_df = crime_df[crime_df["Crime_Status_Desc"] != "UNK"]
crime_df = crime_df.reset_index(drop=True)

In [373]:
crime_df["Crime_Status_Desc"] = (crime_df["Crime_Status_Desc"]
                                .map({"Invest Cont": "Investigation Ongoing",
                                      "Juv Arrest": "Juvenile Arrested",
                                      "Adult Arrest": "Adult Arrested",
                                      "Juv Other": "Juvenile Other"}))

### Checking Datatypes

The Latitude and Longitude columns seem to be fine already. Now I'll check the dtypes to see if any of them can be changed.

In [376]:
crime_df.dtypes

Date_Reported            datetime64[ns]
Date_Occurred            datetime64[ns]
Time_Occurred                    object
Area_Name                        object
District_Number                   int64
Crime_Severity                   object
Crime_Desc                       object
Victim_Age                      float64
Victim_Gender                    object
Victim_Ethnicity                 object
Type_of_Premises_Desc            object
Weapon_Desc                      object
Crime_Status_Desc                object
Latitude                        float64
Longitude                       float64
dtype: object

Victim_Age can be changed from a float to an integer datatype to slightly optimise the data.

In [378]:
crime_df["Victim_Age"] = crime_df["Victim_Age"].astype("Int64")

The datatypes now can't be improved any further without impacting readability of the data. So null values have been dealt with, unnecessary columns removed, and each column has been dealt with one by one to ensure they are consitent and don't have any invalid values. Categorical data has been grouped where necessary so that there is sufficient data to analyse each category. This means the data has now been fully cleaned and is ready for EDA and further analysis in Tableau.

In [380]:
crime_df.head()

Date_Reported Date_Occurred Time_Occurred   Area_Name  District_Number  \
0    2020-03-01    2020-03-01      21:30:00    Wilshire              784   
1    2020-02-09    2020-02-08      18:00:00     Central              182   
2    2020-11-11    2020-11-04      17:00:00   Southwest              356   
3    2023-05-10    2020-03-10      20:37:00    Van Nuys              964   
4    2020-09-09    2020-09-09           NaT  Hollenbeck              413   

  Crime_Severity             Crime_Desc  Victim_Age Victim_Gender  \
0         Severe         Stolen Vehicle        <NA>          Male   
1         Severe  Burglary From Vehicle          47          Male   
2         Severe            Stolen Bike          19           NaN   
3         Severe      Grand Shoplifting          19          Male   
4         Severe         Stolen Vehicle        <NA>           NaN   

  Victim_Ethnicity Type_of_Premises_Desc Weapon_Desc      Crime_Status_Desc  \
0            Other                Street         NaN         Adult Arrested   
1            Other                 Other         NaN  Investigation Ongoing   
2          Unknown   Multi-Unit Dwelling         NaN  Investigation Ongoing   
3            Other        Clothing Store         NaN  Investigation Ongoing   
4          Unknown                Street         NaN  Investigation Ongoing   

   Latitude  Longitude  
0   34.0375  -118.3506  
1   34.0444  -118.2628  
2   34.0210  -118.3002  
3   34.1576  -118.4387  
4   34.0820  -118.2130

# Exporting the Data

In [382]:
crime_df.to_csv("Crime Data Cleaned.csv")